Rs/logging example (#796)
neuralmagic/deepsparse: examples/logging-prometheus/README.md

# DeepSparse + Prometheus/Grafana

This is a simple example that shows you how to connect DeepSparse Logging to the Prometheus/Grafana stack.

#### There are four steps:

> - Configure DeepSparse Logging to log metrics in Prometheus format to a REST endpoint

Reviewer: Not very well versed here, but are we for sure using REST? The fact that we are using HTTP API does not automatically imply REST. Not sure.

> - Point Prometheus to the appropriate endpoint to scape the data at a specified interval

Reviewer: ```suggestion
- Point Prometheus to the appropriate endpoint to scrape the data at a specified interval
```

- DeepSparse Server is installed via PyPi (`pip install deepsparse[server]`)

#### Code
The repository contains all the code you need:

```bash
.
├── client 
│   ├── client.py           # simple client application for interacting with Server
│   └── goldfish.jpeg       # photo of a goldfish
|   └── all_black.jpeg      # photo with just black pixels
├── server-config.yaml      # specifies the configuration of the DeepSparse server
├── custom-fn.py            # custom function used for the logging
├── docker                  # specifies the configuration of the containerized Prometheus/Grafana stack
│   ├── docker-compose.yaml
│   └── prometheus.yaml
└── grafana                 # specifies the design of the Grafana dashboard
    └── dashboard.json
```
## 1. Spin up the DeepSparse Server

`server-config.yaml` specifies the config of the DeepSparse Server, including for logging:

```yaml
# server-config.yaml

loggers:
  prometheus:     # logs to prometheus on port 6100
    port: 6100

endpoints:
  - task: image_classification
    route: /image_classification/predict
    model: zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenet/pruned95_quant-none
    data_logging:      

>       pipeline_inputs.images[0]: # applies to the first image (of the form stage.property[idx])

Reviewer: nit: 
```suggestion
      pipeline_inputs.images[0]: # applies to the first image (of the form target.property[idx])
```
(not married to this naming, but stage sounds odd to me)

```yaml    
# docker-compose.yaml
    
version: "3"

services:
  prometheus:
    image: prom/prometheus
    extra_hosts:
      - "host.docker.internal:host-gateway"     # allow a direct connection from container to the local machine
    ports:
      - "9090:9090" # the default port used by Prometheus
    volumes:
      - ${PWD}/prometheus.yaml:/etc/prometheus/prometheus.yml # mount Prometheus config file

  grafana:
    image: grafana/grafana:latest
    depends_on:
      - prometheus
    ports:
      - "3000:3000" # the default port used by Grafana

```

Prometheus Config file:

```yaml
# prometheus.yaml
    
global:
  scrape_interval: 15s                      # how often to scrape from endpoint
  evaluation_interval: 30s                  # time between each evaluation of Prometheus' alerting rules

scrape_configs:
  - job_name: prometheus_logs               # your project name
    static_configs:
      - targets:
          - 'host.docker.internal:6100'     # should match the port exposed by the PrometheusLogger in the DeepSparse Server config file 
```
</details>

To start up a Prometheus stack to monitor the DeepSparse Server, run:

```bash
cd docker
docker-compose up
```

> ## 3. Launch the Python Client and Run Inference

Reviewer 2: This is an awesome section! great idea

    route: /image_classification/predict
    model: zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenet/pruned95_quant-none
    data_logging:      
      pipeline_inputs.images[0]: # applies to the first image (of the form target.property[idx])
        - func: fraction_zeros   # built-in function
          frequency: 1
          target_loggers:
            - prometheus
        - func: custom-fn.py:mean_pixel_red # custom function
          frequency: 1
          target_loggers:
            - prometheus
```

The config file instructs the server to create an image classification pipeline. Prometheus logs are declared to be exposed on port `6100`, system logging is turned on, and we will log the mean pixel of the red channel (a custom function) as well as the percentage of pixels that are 0 (a built-in function) for each image sent to the server.

Thus, once launched, the Server exposes two endpoints:
- port `6100`: exposes the `metrics` endpoint through [Prometheus python client](https://github.com/prometheus/client_python).
- port `5543`: exposes the endpoint for inference.

To spin up the Server execute:
```
deepsparse.server --config_file server-config.yaml
```

> To validate that metrics are being properly exposed, visit `localhost:6100`. It should contain logs in the specific format meant to be used by the PromQL query language.

Reviewer 3: ```suggestion
To validate that metrics are being properly exposed, visit `localhost:6100`. It should contain logs in the specific format meant to be used by the Prometheus Query Language (PromQL).
```

scrape_configs:
  - job_name: prometheus_logs               # your project name
    static_configs:
      - targets:
          - 'host.docker.internal:6100'     # should match the port exposed by the PrometheusLogger in the DeepSparse Server config file 
```
</details>

To start up a Prometheus stack to monitor the DeepSparse Server, run:

```bash
cd docker
docker-compose up
```

## 3. Launch the Python Client and Run Inference

`client.py` is a simple client that simulates the behavior of some application. In the example, we have two images:
  - `goldfish.jpeg`: sample photo of a Goldfish
  - `all-black.jpeg`: a photo that is all black (every pixel is a 0)  

The client sends requests to the Server, initially with "just" the Goldfish. Over time, we start to randomly
send the All Black image to the server with increasing probability. This simulates a data issue in the 
pipeline that we can detect with the monitoring system.

Run the following to start inference:

```bash
python client/client.py
```

It prints out which image was sent to the server.

## 4. Inspecting the Prometheus/Grafana Stack

### Prometheus

> #### Confirm It Is Working

Reviewer 3: ```suggestion
#### Confirming Prometheus is Working
```

> Visiting `http://localhost:9090/targets` should show that an endpoint `http://host.docker.internal:6100/metrics` is in state `UP`.

Reviewer 3: ```suggestion
Visiting `http://localhost:9090/targets` should show that an endpoint `http://host.docker.internal:6100/metrics` is in the `UP` state.
```

> #### Query Prometheus with PromQL

Reviewer 3: ```suggestion
#### Querying Prometheus with PromQL
```

> If you do not want to use Grafana, you can start off by using Prometheus's native graphing functionality. 

Reviewer 3: ```suggestion
If you do not want to use Grafana, you can use Prometheus's native graphing functionality. 
```

> 

Reviewer 3: COMBINE THESE TWO PARAGRAPHS

Navigate to `http://localhost:9090/graph` and add the following `Expression`:

```
rate(image_classification__0__pipeline_inputs__images__fraction_zeros_sum[30s])
/
rate(image_classification__0__pipeline_inputs__images__fraction_zeros_count[30s])
```

> You should see the following:

Reviewer 3: ```suggestion
You should see:
```

![prometheus-dashboard.png](image/prometheus-dashboard.png)

> This graph shows the percentage of 0 pixels in the images sent to the server. 

Reviewer 3: ```suggestion
This graph shows the percentage of 0 pixels in the images sent to the Server. 
```

> As the "corrupted" all black images were sent to the server in increasing probability, 

Reviewer 3: ```suggestion
As the "corrupted" all-black images were sent to the Server in increasing probability, 
```

we can clearly see a spike in the graph, alerting us
that something strange is happening with the provided input.

DeepSparse Server also automatically logs prediction latencies for each pipeline stage as well
end-to-end server-side inference time. Add the following query to inspect average latency:

```
rate(image_classification__0__prediction_latency__total_inference_sum[30s])
/
rate(image_classification__0__prediction_latency__total_inference_count[30s])
```

![prometheus-dashboard-latency.png](image/prometheus-dashboard-latency.png)

> For more details on working with the Prometheus Query Language PromQL, 

Reviewer 3: ```suggestion
```

> see [the official docs](https://prometheus.io/docs/prometheus/latest/querying/basics/).

Reviewer 3: ```suggestion
See [Querying Prometheus](https://prometheus.io/docs/prometheus/latest/querying/basics/) for more details on working with PromQL.
```

### Grafana

> #### Login

Reviewer 3: ```suggestion
#### Logging In
```

Visit `localhost:3000` to launch Grafana. Log in with the default username (`admin`) and password (`admin`). 

> #### Add Prometheus Data Source

Reviewer 3: ```suggestion
#### Adding Prometheus Data Source
```

> Setup the Prometheus data source (`Add your first data source` -> `Prometheus`). On this page, we just

Reviewer 3: ```suggestion
Set up the Prometheus data source (`Add your first data source` -> `Prometheus`). On this page, we just
```

need to update the `url` section. Since Grafana and Prometheus are running separate docker containers, 

> put we need to put the IP address of the Prometheus container.

Reviewer 3: ```suggestion
we need to include the IP address of the Prometheus container.
```

> Run the following to lookup the `name` of your Prometheus container:

Reviewer 3: ```suggestion
Run the following to look up the `name` of your Prometheus container:
```

```
docker container ls
>>> CONTAINER ID   IMAGE                    COMMAND                  CREATED             STATUS             PORTS                                       NAMES
>>> 997521854d84   grafana/grafana:latest   "/run.sh"                About an hour ago   Up About an hour   0.0.0.0:3000->3000/tcp, :::3000->3000/tcp   docker_grafana_1
>>> c611c80ae05e   prom/prometheus          "/bin/prometheus --c…"   About an hour ago   Up About an hour   0.0.0.0:9090->9090/tcp, :::9090->9090/tcp   docker_prometheus_1
```

> Run the following to lookup the IP address (replace `docker_prometheus_1` with your container's name):

Reviewer 3: ```suggestion
Run the following to look up the IP address (replace `docker_prometheus_1` with your container's name):
```

```
docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' docker_prometheus_1
>>> 172.18.0.2
```

> So, in our case, the `url` section should be: `http://172.18.0.2:9090`.

Reviewer 3: ```suggestion
So, in our case, the `url` section should be `http://172.18.0.2:9090`.
```

> Click `Save & Test`. We should get a green check saying "Data Source Is Working".

Reviewer 3: ```suggestion
Click `Save & Test`. We should get a green check indicating, "Data Source Is Working."
```

> #### Import A Dashboard

Reviewer 3: ```suggestion
#### Importing a Dashboard
```

Now you should be ready to create/import your dashboard. 

> Grafana's interface for adding metrics is very intuitive (and you can use PromQL), 

Reviewer 3: ```suggestion
Although Grafana's interface for adding metrics is very intuitive (and you can use PromQL), 
```

> but we have provided a simple pre-made dashboard for this use case. 

Reviewer 3: ```suggestion
we have provided a simple pre-made dashboard for this use case. 
```

> Click `Dashboard` -> `Import` on the left-hand side bar. You should see an option to upload a file.

Reviewer 3: ```suggestion
Click `Dashboard` -> `Import` on the left-hand sidebar. You should see an option to upload a file.
```
